pkg/challenge/challenge.go

```diff
@@ -0,0 +1,173 @@
+package challenge
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/rand"
+	"fmt"
+	"time"
+
+	"github.com/kolide/krypto"
+	"github.com/kolide/krypto/pkg/echelper"
+	"github.com/vmihailenco/msgpack/v5"
+	"golang.org/x/crypto/nacl/box"
+)
+
+type OuterChallenge struct {
+	Sig            []byte `msgpack:"sig"`
+	Msg            []byte `msgpack:"msg"`
+	innerChallenge *InnerChallenge
+}
+
+func (o *OuterChallenge) Verify(counterParty ecdsa.PublicKey) error {
+	if err := echelper.VerifySignature(counterParty, o.Msg, o.Sig); err != nil {
+		return err
+	}
+
+	innerChallenge, err := o.inner()
+	if err != nil {
+		return err
+	}
+
+	o.innerChallenge = innerChallenge
+	return nil
+}
+
+func (o *OuterChallenge) RequestData() []byte {
+	if o.innerChallenge != nil {
+		return o.innerChallenge.RequestData
+	}
+
+	return nil
+}
+
+func (o *OuterChallenge) TimeStamp() int64 {
```

Comment:
I think of timestamp as one word. `Timestamp`

Reply:
updated